node/src/components/chain_synchronizer/operations.rs

```diff
@@ -341,125 +347,196 @@ impl CanUseSyncingNodes for Deploy {}
 }
 
-/// Returns fully-connected, syncing and non-syncing peers that are known to be not banned.
-async fn get_filtered_fully_connected_peers<REv>(ctx: &ChainSyncContext<'_, REv>) -> Vec<NodeId>
+/// Fetching should stop only when the network is connected and all retries have been exhausted.
+fn should_stop_fetching(
+    retry_count: &mut usize,
+    has_connected_to_network: bool,
+    max_sync_fetch_attempts: usize,
+) -> bool {
+    if !has_connected_to_network {
+        return false;
+    }
+
+    *retry_count += 1;
+    if *retry_count <= max_sync_fetch_attempts {
+        return false;
+    }
+
+    true
+}
+
+#[derive(Error, Debug)]
+pub(crate) enum FetchWithRetryError<T>
 where
-    REv: From<NetworkInfoRequest>,
+    T: Item,
 {
-    let mut peer_list = ctx.effect_builder.get_fully_connected_peers().await;
-    ctx.filter_bad_peers(&mut peer_list);
-    peer_list
+    #[error("Fetch retries exhausted for item with id {id:?}. Total attempts: {total_attempts}, retries while network was connected: {retries_while_connected}")]
+    RetriesExhausted {
+        id: T::Id,
+        total_attempts: usize,
+        retries_while_connected: usize,
+    },
+
+    #[error(transparent)]
+    FetcherError(#[from] FetcherError<T>),
 }
 
-/// Fetches an item. Keeps retrying to fetch until it is successful. Not suited to fetching a block
-/// header or block by height, which require verification with finality signatures.
-async fn fetch_retry_forever<REv, T>(ctx: &ChainSyncContext<'_, REv>, id: T::Id) -> FetchResult<T>
+/// Fetches an item. Not suited to fetching a block header or block by height, which require
+/// verification with finality signatures.
+async fn fetch_with_retries<REv, T>(
+    ctx: &ChainSyncContext<'_, REv>,
+    id: T::Id,
+) -> Result<FetchedData<T>, FetchWithRetryError<T>>
 where
     T: Item + CanUseSyncingNodes + 'static,
     REv: From<FetcherRequest<T>> + From<NetworkInfoRequest>,
 {
-    let mut attempts = 0_usize;
+    let mut total_attempts = 0_usize;
+    let mut retry_count = 0_usize;
     loop {
-        let new_peer_list = if T::can_use_syncing_nodes() {
-            get_filtered_fully_connected_peers(ctx).await
-        } else {
-            get_filtered_fully_connected_non_syncing_peers(ctx).await
-        };
+        let has_connected_to_network = has_connected_to_network();
+        if should_stop_fetching(
+            &mut retry_count,
+            has_connected_to_network,
+            ctx.config.max_sync_fetch_attempts(),
+        ) {
+            return Err(FetchWithRetryError::RetriesExhausted {
+                id,
+                total_attempts,
+                retries_while_connected: retry_count - 1,
+            });
+        }
 
-        if new_peer_list.is_empty() && attempts % 100 == 0 {
-            warn!(
-                attempts,
+        let new_peer_list = get_peers(T::can_use_syncing_nodes(), ctx).await;
+        if new_peer_list.is_empty() && total_attempts % 100 == 0 {
+            error!(
```

Comment:
Shouldn't this be a `warn!`?  Also, might be worthwhile to include the `retry_count` in there.

Reply:
Yes it should, I promoted it to `error` to eagerly. Reverted and included the `retry_count` in 7a0a181f2a40f80e0757ef65476d4a3c49f7224a